### ISSUE TITLE
Unable to create AAD User

### ISSUE THREAD

**Issue by shalinin00** [Azure/azure-service-operator#3344]
**Version of Azure Service Operator**
ASO V2 being used to create flexibleserversadministrator  and user_aad. Facing below issue - 

```
0927 03:48:29.053859       1 aad_user.go:134] controllers/UserController "msg"="Retrieved token for MySQL" "azureName"="testaadusers" "expires"="2023-09-28T01:28:57.934530917Z" "name"="testaadusers" "scope"="https://ossrdbms-aad.database.windows.net/.default"
E0927 03:48:29.088605       1 generic_reconciler.go:360] controllers/UserController "msg"="Encountered error impacting Ready condition" "error"="Reason: Failed, Severity: Warning, RetryClassification: RetrySlow, Cause: failed to connect database. Server: asotest-mysql-server.mysql.database.azure.com, Database: mysql, Port: 3306, AdminUser: shalinin: error pinging the mysql db (asotest-mysql-server.mysql.database.azure.com:3306/mysql): Error 1045: Access denied for user 'shalinin@'51.105.253.159' (using password: YES)" "name"="testaadusers" 

```

I have already created flexibleserver and flexibleserversdatabase by following this wiki - https://github.com/Azure/azure-service-operator/blob/09b0cef5265ce6c6be7eee4643508037de6e43d0/v2/samples/dbformysql/v1api/v1api20210501_flexibleserver.yaml

Managed identity and federated identity created as well using this link - https://github.com/Azure/azure-service-operator/blob/09b0cef5265ce6c6be7eee4643508037de6e43d0/v2/samples/managedidentity/v1api/v1api20181130_userassignedidentity.yaml

**Describe the bug**
Create flexibleserversadministrator  and facing issue in creating user_aad.yaml and user.yaml. Facing below issue for user_aad.yaml- 

```
0927 03:48:29.053859       1 aad_user.go:134] controllers/UserController "msg"="Retrieved token for MySQL" "azureName"="testaadusers" "expires"="2023-09-28T01:28:57.934530917Z" "name"="testaadusers" "scope"="https://ossrdbms-aad.database.windows.net/.default"
E0927 03:48:29.088605       1 generic_reconciler.go:360] controllers/UserController "msg"="Encountered error impacting Ready condition" "error"="Reason: Failed, Severity: Warning, RetryClassification: RetrySlow, Cause: failed to connect database. Server: asotest-mysql-server.mysql.database.azure.com, Database: mysql, Port: 3306, AdminUser: shalinin: error pinging the mysql db (asotest-mysql-server.mysql.database.azure.com:3306/mysql): Error 1045: Access denied for user 'shalinin@adobe.com'@'51.105.253.159' (using password: YES)" "name"="testaadusers" 

```
**To Reproduce**
1. Followed the steps - https://github.com/Azure/azure-service-operator/blob/09b0cef5265ce6c6be7eee4643508037de6e43d0/v2/samples/dbformysql/v1api/v1_user.yaml
2.  Admin created well - 
```
kubectl get flexibleserversadministrator.dbformysql.azure.com/shalinin
NAME       READY   SEVERITY   REASON      MESSAGE
shalinin   True               Succeeded

```
3. Facing issue when created user_aad.yaml - 

```
kubectl get user.dbformysql.azure.com/testaadusers
NAME           READY   SEVERITY   REASON   MESSAGE
testaadusers   False   Warning    Failed   failed to connect database. Server: asotest-mysql-server.mysql.database.azure.com, Database: mysql, Port: 3306, AdminUser: shalinin: error pinging the mysql db (asotest-mysql-server.mysql.database.azure.com:3306/mysql): Error 1045: Access denied for user 'shalinin'@'51.123.123.159' (using password: YES)
```

**Expected behavior**
Should be able to connect to the server without providing password as using managed identity. 

**Screenshots**

**Additional context**
aad_user.yaml file - 

```
apiVersion: dbformysql.azure.com/v1
kind: User
metadata:
  name: testaadusers
  namespace: abc
spec:
  owner:
    name: asotest-mysql-server
  # Specify a list of server-level privileges. Privileges
  # include the following: RELOAD, PROCESS, SHOW DATABASES,
  # REPLICATION SLAVE, REPLICATION CLIENT, CREATE USER
  privileges:
    - PROCESS
    - CREATE USER
  databasePrivileges:
    mysqldatabase-sample:
      # Privileges include the following:
      # SELECT, INSERT, UPDATE, DELETE, CREATE, DROP, REFERENCES, INDEX,
      # ALTER, CREATE TEMPORARY TABLES, LOCK TABLES, EXECUTE, CREATE VIEW,
      # SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, EVENT, TRIGGER
      - SELECT
  aadUser:
    # alias is optional if the name of the user is less than 32 characters but required if it's longer
    # than 32 characters. Note that if alias is specified it is the name used to configure/control the user
    # via SQL commands, but the actual name you use to log into the Server with is the name specified in .spec.azureName.
    # For example if my azureName is "myuser@mydomain.onmicrosoft.com" and alias is "myuser", when logging in specify
    # username "myuser@mydomain.onmicrosoft.com".
    #alias: myalias
    # The serverAdminUsername should match the login name of the server AAD administrator.
    # If the administrator was configured with ASO, that field is .spec.login on the FlexibleServersAdministrator type.
    # If the administrator is a group, the ServerAdminUsername should be the group name, not the actual username of the
    # identity to log in with. For example if the administrator group is "admin-group" and identity "my-identity" is
    # a member of that group, the ServerAdminUsername should be "admin-group"
    serverAdminUsername: "shalinin"

```

**Comment by theunrepentantgeek** [comment]
I think this error message is the key:

```
$ kubectl get user.dbformysql.azure.com/testaadusers
NAME           READY   SEVERITY   REASON   MESSAGE
testaadusers   False   Warning    Failed   failed to connect database. Server: asotest-mysql-server.mysql.database.azure.com, Database: mysql, Port: 3306, AdminUser: shalinin: error pinging the mysql db (asotest-mysql-server.mysql.database.azure.com:3306/mysql): Error 1045: Access denied for user 'shalinin'@'51.123.123.159' (using password: YES)
```

You've created a resource in your cluster requesting ASO create the user `testaadusers` in your MySQL database - but ASO is getting an access denied error trying to do that.

The `(using password: YES)` part leads me to think the AAD connection isn't working, and the MySQL server is expecting the admin username and password to be provided.

The error is most likely in your [`FlexibleServersAdministrator`](https://azure.github.io/azure-service-operator/reference/dbformysql/v1api20220101/#dbformysql.azure.com/v1api20220101.FlexibleServersAdministrator) instance - please share that yaml (redact the `sid` property as that's sensitive).

Also, I suspect (but haven't verified) that the `administratorLogin` given on the server must be different to the AAD user name.

**Comment by shalinin00** [comment]
Thank you so much for response. 
Now, im getting this error -
```
kubectl get user.dbformysql.azure.com/testaadusers

NAME           READY   SEVERITY   REASON   MESSAGE
testaadusers   False   Warning    Failed   failed to get token from credential: WorkloadIdentityCredential authentication failed...
```

Now, my resource, managed identity and federated identity is failing with the same error - saying 
```
kubectl get resourcegroup.resources.azure.com/aso-rg
NAME     READY   SEVERITY   REASON   MESSAGE
aso-rg   False   Warning    Failed    WorkloadIdentityCredential authentication failed.
```..

Below is my flexibleserver, flexibleserveradministartor and aad_user.yaml attached.
[temp.zip](https://github.com/Azure/azure-service-operator/files/12754627/temp.zip)

**Comment by shalinin00** [comment]
```
kubectl describe resourcegroup.resources.azure.com/aso-rg


message:               WorkloadIdentityCredential authentication failed
POST https://login.microsoftonline.com/
--------------------------------------------------------------------------------
RESPONSE 400 Bad Request
--------------------------------------------------------------------------------
{
  "error": "unauthorized_client",
  "error_description": "AADSTS700016: Application with identifier 'https://abc' was not found in the directory 'X'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.\r\nTrace ID: \r\nCorrelation ID: \r\nTimestamp: 2023-09-28 23:59:03Z”,} 
```

**Comment by matthchr** [comment]
I suspect the original problem you were having is due to this:
https://learn.microsoft.com/en-us/azure/mysql/flexible-server/how-to-azure-ad#grant-permissions-to-user-assigned-managed-identity.

Specifically, you must grant the following graph permissions to the identity which is specified in the `identity` section of the FlexibleServer:
User.Read.All, GroupMember.Read.All, and Application.Read.ALL

Of course... we don't document that anywhere in our samples which is a big gap.

On to the next error, the workload identity error: This "Application with identifier 'https://abc' was not found in the directory 'X'" makes me think that the UMI was deleted somehow? Or possibly the federated identity credential on the UMI was? Can you check and confirm in the Azure portal (and/or ASO by examining those two resources) that the UMI and FederatedIdentityCredential still exist?

**Comment by matthchr** [comment]
> Of course... we don't document that anywhere in our samples which is a big gap.

I'll take an action item from this issue to fix up our samples to make this much clearer.

**Comment by matthchr** [comment]
See #3349 which contains some documentation improvments that hopefully make this more understandable.

**Comment by shalinin00** [comment]
For  `WorkloadIdentityCredential authentication failed `- This is very strange. I think the problem is when i delete the resources which i dont want then this error starts coming up in all the other resources, UMI, Federated credentials. 

```
kubectl get resourcegroup.resources.azure.com
 
NAME                READY   SEVERITY   REASON         MESSAGE
aso-rg-sample       True               Succeeded
aso-rg-test         True               Succeeded
aso-rgsample-test   True               Succeeded
aso-test            True               Succeeded
mysql-rg            False   Warning    UnknownError   WorkloadIdentityCredential authentication failed...
mysqltest-rg        False   Warning    UnknownError   WorkloadIdentityCredential authentication failed...
```
I suspect that when you delete a resource and that doesnt get deleted quickly and takes time then everything which are present start showing such errors and along with this - 

```
message:               WorkloadIdentityCredential authentication failed
POST https://login.microsoftonline.com/
--------------------------------------------------------------------------------
RESPONSE 400 Bad Request
--------------------------------------------------------------------------------
{
  "error": "unauthorized_client",
  "error_description": "AADSTS700016: Application with identifier 'https://abc' was not found in the directory 'X'. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You may have sent your authentication request to the wrong tenant.\r\nTrace ID: \r\nCorrelation ID: \r\nTimestamp: 2023-09-28 23:59:03Z”,} 

```

**Comment by shalinin00** [comment]
Also, i just noticed that,  when i created flexibleserver named 'asotest-mysql-server' and **administratorLogin** as shalinin and then tried to create **flexibleserveradministartor**, it got **created**. 
However,  when i created another flexibleserver named 'asotest-mysql-server-one' and **administratorLogin** as shalinin then it complained - 

```
kubectl apply -f flexibleserveradministrator.yaml

Error from server (Forbidden): error when applying patch:
{"metadata":{"annotations":{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"dbformysql.azure.com/v1api20220101\",\"kind\":\"FlexibleServersAdministrator\",\"metadata\":{\"annotations\":{},\"name\":\"shalinin\",\"namespace\":\"preview\"},\"spec\":{\"administratorType\":\"ActiveDirectory\",\"identityResourceReference\":{\"group\":\"managedidentity.azure.com\",\"kind\":\"UserAssignedIdentity\",\"name\":\"testuserassignedidentity\"},\"login\":\"testuserassignedidentity\",\"owner\":{\"name\":\"asotest-mysql-server-one\"},\"sidFromConfig\":{\"key\":\"clientId\",\"name\":\"admin-settings\"},\"tenantIdFromConfig\":{\"key\":\"tenantId\",\"name\":\"admin-settings\"}}}\n"}},"spec":{"owner":{"name":"asotest-mysql-server-one"}}}
to:
Resource: "dbformysql.azure.com/v1api20220101, Resource=flexibleserversadministrators", GroupVersionKind: "dbformysql.azure.com/v1api20220101, Kind=FlexibleServersAdministrator"
Name: "shalinin", Namespace: "preview"
for: "flexibleserveradministrator.yaml": error when patching "flexibleserveradministrator.yaml": admission webhook "validate.v1api20220101.flexibleserversadministrators.dbformysql.azure.com" denied the request: updating 'Owner.Name' is not allowed for 'dbformysql.azure.com/v1api20220101, Kind=FlexibleServersAdministrator : shalinin
```

Just want to understand here, if we create flexibleserversadministrators named shalinin then we can not use it in creating another flexibleserver. Here, only one time flexibleserversadministrators = shalinin be created.

**Comment by shalinin00** [comment]
> I suspect the original problem you were having is due to this: https://learn.microsoft.com/en-us/azure/mysql/flexible-server/how-to-azure-ad#grant-permissions-to-user-assigned-managed-identity.
> 
> Specifically, you must grant the following graph permissions to the identity which is specified in the `identity` section of the FlexibleServer: User.Read.All, GroupMember.Read.All, and Application.Read.ALL
> 
Yes, thats right, I believe I dont have admin permission granted to UMI. Thank you, this really helps!!

**Comment by matthchr** [comment]
You had a couple quesitons:

## Question 1: 
> This is very strange. I think the problem is when i delete any of existing resources which i dont want then this error starts coming up in all the other resources, UMI, Federated credentials.

How and what resources did you delete?

Part of the instruction for creating AAD MySQL Users is to:
1. Create a Managed Identity
2. Use that managed identity as ASOs identity.

So if you deleted the managed identity which ASO was using as its managed identity (the identity which it is communicating to Azure with), then it's expected that all of its calls to Azure fail.

If you don't want to use that identity for the whole operator and instead only want to use it for managing the SQL user, you can look into the [credential scope](https://azure.github.io/azure-service-operator/guide/authentication/credential-scope/#resource-scope) options, and specify `serviceoperator.azure.com/credential-from` on the `User` object. Or based on the snippet you've shown maybe you're already using that but you deleted things in the wrong order, if you're using that open to make ASO use the UMI for the MySQL RGs or namespaces, and you delete the UMI, then ASO will be unable to manage those resources anymore.

You need to delete the resources in ASO first, and then optionally the UMI. (If you're using namespace scoped secrets really you should delete the whole namespace, and only then delete the UMI for that namespace)

## Question 2:

> Just want to understand here, if we create flexibleserversadministrators named shalinin then we can not use it in creating another flexibleserver. As shalinin is now bounded with 1 flexible server. If tries to connect shalinin with 2 flexible server, it is unable to connect and throw above patch error. 

It looks to me like what you did here is tried to _edit_ an existing ASO FlexibleServerAdministrator object named `shalinin` and have it point to a different FlexibleServer owner (at least, that's what the webhook is saying). You cannot do this.

Possibly the confusion here is around "name in Kubernetes" versus "name in Azure".

You _can_ create a different FlexibleServersAdministrator, with a different name in Kubernetes (say `shalinin2`) and have that point to the new flexible server. It can even (I think) share the same name (in Azure), by setting the `azureName` field in the CRD to whatever you like. This is because name-uniqueness in Azure is the "fully qualified name" (flexibleServers/foo/administrators/shalinin is different than flexibleServers/bar/administrators/shalinin), but in Kubernetes its by Namespace + Name.

Basically there is a 1-1 mapping between "resource in Kubernetes" and "resource in Azure". Since you've got 2 flexible servers, with with different administrators (the admins may be the same identity, but the actual administrator object is 1 per flexible server) you need 2 Kubernetes resources to match, not 1.

## Question 3:

I just created flexibleserversadministrators named with my UMI and getting error -InternalServerError   An unexpected error occured while processing the request. Tracking ID: '87903e': GET https://management.azure.com/subscriptions/98790/providers/Microsoft.DBforMySQL/locations/westeurope/azureAsyncOperation/987680...

This _looks_ like an error from the MySQL RP itself. We might need to engage them on this if it keeps happening consistently.

**Comment by shalinin00** [comment]
> using that open to make ASO use the UMI for the MySQL RGs or namespaces, and you delete the UMI, then ASO will be unable to manage those resources anymore.

Thank you so much for details here and quick turnover.  
Yes, i deleted the MySQL RGs that was using the UMI for ASO which make sense. Generally, when i tried to delete resources, it takes forever to delete, maybe its being attached with the UMI for making ASO call.

**Comment by shalinin00** [comment]
> Basically there is a 1-1 mapping between "resource in Kubernetes" and "resource in Azure". Since you've got 2 flexible servers, with with different administrators (the admins may be the same identity, but the actual administrator object is 1 per flexible server) you need 2 Kubernetes resources to match, not 1.

Thank you so much, understood, now make sense. Let me try it out and let you know.

**Comment by matthchr** [comment]
Can you share the YAML for `testuserassignedidentity`?

The [sample we have](https://github.com/Azure/azure-service-operator/blob/main/v2/samples/dbformysql/v1api/v1api20220101_flexibleserversadministrator.yaml) shows using a UMI as the administrator identity.

My assumption is that the `InternalServerError` you're seeing is because of some wrong configuration on the client side. Once we get a chance to look at that hopefully we can help you to figure out what is the problem.

**Comment by shalinin00** [comment]
> https://github.com/Azure/azure-service-operator/files/12754631/temp.zip

I use this as managed identity - 
```
apiVersion: managedidentity.azure.com/v1api20181130
kind: UserAssignedIdentity
metadata:
  name: testUMI
  namespace: abc
spec:
  location: us
  owner:
    name: aso-rg-test
```

**Comment by shalinin00** [comment]
When I use operatorspec then i get the following error - 

```
apiVersion: managedidentity.azure.com/v1api20181130
kind: UserAssignedIdentity
metadata:
  name: testUMI
  namespace: abc
spec:
  location: us
  owner:
    name: aso-rg-test
operatorSpec:
    configMaps:
      tenantId:
        name: admin-settings
        key: tenantId
      clientId:
        name: admin-settings
        key: clientId
```

`Error from server (BadRequest): error when creating "testUMI": UserAssignedIdentity in version "v1api20181130" cannot be handled as a UserAssignedIdentity: strict decoding error: unknown field "operatorSpec"`

**Comment by shalinin00** [comment]
@matthchr Any thoughts on the above error?

**Comment by theunrepentantgeek** [comment]
YAML is whitespace sensitive. 

Assuming the YAML you've quoted above is accurate, check the indentation of `operatorSpec` - you have it as a property at the same level as `metadata` and `spec`, which would correctly generate an error because there is no `operatorSpec` property at that level.

Indent it to the same level as `location` and `owner` and it's a different story - because there is an `operatorSpec` property within `spec`.

**Comment by shalinin00** [comment]
@theunrepentantgeek @matthchr 
Thank you, i was able to resolve this. However, I still can not create AAD user. 

```
apiVersion: dbformysql.azure.com/v1
kind: User
metadata:
  name: testadministratoruser
  namespace: abc
spec:
  owner:
    name: abc
  privileges:
    - PROCESS
    - CREATE USER
  databasePrivileges:
    testmydb:
      - SELECT
  aadUser:
    serverAdminUsername: testumi
```

```
apiVersion: dbformysql.azure.com/v1api20210501
kind: FlexibleServersDatabase
metadata:
  name: testmydb
  namespace: abc
spec:
  owner:
    name: abc
  charset: utf8mb4
```

**Comment by shalinin00** [comment]
@theunrepentantgeek @matthchr 
Also, i cannot create **user** - 

```
apiVersion: dbformysql.azure.com/v1
kind: User
metadata:
  name: testuserwithumi
  namespace: abc
spec:
  owner:
    name: abc
  # Specify a list of server-level privileges. Privileges
  # include the following: RELOAD, PROCESS, SHOW DATABASES,
  # REPLICATION SLAVE, REPLICATION CLIENT, CREATE USER
  privileges:
    - PROCESS
    - CREATE USER
    - SHOW DATABASES
  databasePrivileges:
    testmydb:
      # Privileges include the following:
      # SELECT, INSERT, UPDATE, DELETE, CREATE, DROP, REFERENCES, INDEX,
      # ALTER, CREATE TEMPORARY TABLES, LOCK TABLES, EXECUTE, CREATE VIEW,
      # SHOW VIEW, CREATE ROUTINE, ALTER ROUTINE, EVENT, TRIGGER
      - SELECT
      - CREATE
      - UPDATE
      - INSERT
  localUser:
    serverAdminUsername: shalinin3
    # serverAdminPassword is optional. If NOT specified, the operator uses its identity to log into the server.
    # The operator can only successfully log into the server if its identity is the administrator of the server or
    # if its identity is a member of a group which is the administrator of the server or if its identity is a member of a
    # group which is the administrator of the server.
    serverAdminPassword:
      name: mysql-secret
      key: PASSWORD
    password:
      name: sampleuser-password
      key: password
```

**Comment by matthchr** [comment]
Have you configured any firewall rules for your database? Or joined it and the cluster ASO is in to the same vnet?

AFAIK, by default, the FlexibleServer will block connections attempts unless a firewall rule exists allowing connections from that endpoint, or if they're in the same vnet.

**Comment by shalinin00** [comment]
@matthchr 
```
Now im getting this error in MID when creating a new one - 
RESPONSE 401 Unauthorized
--------------------------------------------------------------------------------
{
  "error": "invalid_client",
  "error_description": "AADSTS70021: No matching federated identity record found for presented assertion. Assertion Issuer: 'https://.azure.com/JSJSJS/UYSH99/'. Assertion Subject: 'system:serviceaccount:azureserviceoperator-system:azureserviceoperator-default'. Assertion Audience: 'api://AzureADTokenExchange'. https://docs.microsoft.com/en-us/azure/active-directory/develop/workload-identity-federation
```

So, i have one service account named azureserviceoperator-system in namespace azureserviceoperator-default. 
Also, service account named azureserviceoperator-system in namespace abc. 
Dont why it is picking up SA from namespace azureserviceoperator-default.

**Comment by theunrepentantgeek** [comment]
Why do both of your Service Accounts have the same name of `azureserviceoperator-system`? 

Even though they're distinct in your cluster (being in different namespaces), I suspect there's a real possibility they will both map to the same underlying Azure resource because the cluster namespace is _not_ a part of what's reconciled with Azure.

**Comment by matthchr** [comment]
> Dont why it is picking up SA from namespace azureserviceoperator-default.

ASO always uses its service account when communicating with Azure. So whenever you set up federated identity credentials for a managed identity, you need to give its Service Account the ability to exchange its service account token for an AAD token. That's what the FederatedIdentityCredential is doing.

I don't know what your other service account is for (or as @theunrepentantgeek said, why it's also named `azureserviceoperator-system`, as that seems a weird name for a service account that isn't ASOs account), but the relationship you need to set up with your identities and federated identity credentials is a many-to-one relationship. Many FICs to one ServiceAccount, the `azureserviceoperator-default` account in the `azureserviceoperator-system` namespace.

**Comment by matthchr** [comment]
Closing this since there has been no response. If you're still having this issue feel free to open a new one or reopen this one.
